SONARSLANG-125 SonarRuby: Basic AST mapping (Java version)

christophe-zurn-sonarsource opened the pull request.

saberduck commented on sonar-ruby-plugin/src/test/java/org/sonarsource/ruby/converter/RubyConverterTest.java:
```diff
@@ -109,28 +91,26 @@ public void invalid_escape_sequence() {
-    assertTree(converter.parse(("true\r\nfalse"))).isInstanceOf(TopLevelTree.class);
-  }
-
-  @Test
```
Looks like this annotation was dropped accidentally, ranges is now not executed as test

christophe-zurn-sonarsource replied:
Nice catch :+1:

saberduck commented on sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyConverter.java:
```diff
@@ -137,12 +139,28 @@ private Tree parseContent(String content) {
+    }
+  }
+
+  private RubyVisitor getRubyProcessor(TreeMetaDataProvider metaDataProvider) {
```
Should we rename this method to `getRubyVisitor` ?

christophe-zurn-sonarsource replied:
Yep, to be consistent I renamed all remaining "rubyProcessor" to "rubyVisitor"

saberduck commented on sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java:
```diff
@@ -0,0 +1,211 @@
+
+  @JRubyMethod
+  public IRubyObject process(ThreadContext context, IRubyObject node) {
+    // some nodes present in the AST but are empty (e.g. function parameters clause)
```
missing verb: some nodes **are** present ...

christophe-zurn-sonarsource replied:
updated comment

saberduck commented on sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java:
```diff
@@ -0,0 +1,211 @@
+    if (processedNode instanceof NodeAdapter) {
+      // A more specific tree was already created by a specialized method (on_${type}, ex: on_int)
+      return processedNode;
+
```
nitpick, remove whitespace

christophe-zurn-sonarsource replied:
done :)

saberduck commented on sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java:
```diff
@@ -0,0 +1,211 @@
+
+  @JRubyMethod(name = "on_class")
+  public IRubyObject onClass(ThreadContext context, IRubyObject node) {
+    IRubyObject updatedNode = callSuperMethod(context, "on_class", node);
```
maybe we can rename `callSuperMethod`  to `continueVisit` , so it is more clear what it's doing

christophe-zurn-sonarsource replied:
Renamed method to `visitNode`

saberduck commented on sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java:
```diff
@@ -0,0 +1,211 @@
+    return convertToNodeAdapter(node, metaData -> new LiteralTreeImpl(metaData, String.valueOf(value)));
+  }
+
+  private NodeAdapter convertToNodeAdapter(IRubyObject node, Function<TreeMetaData, Tree> getTree) {
```
Can this be static factory method in `NodeAdapter` directly?

christophe-zurn-sonarsource replied:
I'll leave this for now as is, and we'll improve that in another PR

saberduck commented on sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java:
```diff
@@ -0,0 +1,211 @@
+    return method.call(context, this, superclass, methodName, arg1);
+  }
+
+  private Object getChild(IRubyObject node, int index) {
```
Can this method be moved to `NodeAdapter` as instance method?

christophe-zurn-sonarsource replied:
I'll leave this for now as is, and we'll improve that in another PR

saberduck commented on sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java:
```diff
@@ -0,0 +1,211 @@
+    return (IRubyObject) JavaEmbedUtils.invokeMethod(getRuntime(), node, "type", null, IRubyObject.class);
+  }
+
+  private List<Tree> getChildren(IRubyObject node, TreeMetaData metaData) {
```
Can this method be moved to NodeAdapter as instance method?

christophe-zurn-sonarsource replied:
I'll leave this for now as is, and we'll improve that in another PR

saberduck commented on sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java:
```diff
@@ -0,0 +1,211 @@
+    return ((List) JavaEmbedUtils.invokeMethod(getRuntime(), node, "to_a", null, List.class)).get(index);
+  }
+
+  private IRubyObject nodeType(IRubyObject node) {
```
Can this method be moved to NodeAdapter as instance method?

christophe-zurn-sonarsource replied:
I'll leave this for now as is, and we'll improve that in another PR

saberduck commented on sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java:
```diff
@@ -0,0 +1,211 @@
+    return false;
+  }
+
+  private TreeMetaData getMetaData(IRubyObject node) {
```
Can this be moved to `NodeAdapter` ?

christophe-zurn-sonarsource replied:
I'll leave this for now as is, and we'll improve that in another PR

saberduck commented on sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/adapter/NodeAdapter.java:
```diff
@@ -0,0 +1,94 @@
+import org.sonarsource.slang.api.Tree;
+
+@JRubyClass(name = "NodeAdapter")
+public class NodeAdapter extends RubyObject {
```
why it doesn't extend `JRubyObject` ?

christophe-zurn-sonarsource replied:
Do you mean `IRubyObject`? `IRubyObject` is an interface, and we don't want to implement all base methods.
`RubyObject` is the equivalent of the ruby `Object` class, so classes that will be used inside the ruby runtime should extend this basic class.

saberduck commented on slang-checks/src/main/java/org/sonarsource/slang/checks/BadFunctionNameCheck.java:
```diff
@@ -19,28 +19,42 @@
+
+  private static final Map<String, String> DEFAULT_BY_LANGUAGE;
+  static {
+    HashMap<String, String> defaults = new HashMap<>();
```
wdyt to have `enum` for languages in the `checks` module and use `EnumMap` here?

christophe-zurn-sonarsource replied:
I'll leave this for now as is, and we'll improve that in another PR

saberduck commented on sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/plugin/KotlinRulesDefinition.java:
```diff
@@ -35,10 +36,15 @@ public void define(Context context) {
     checks.add(CommentedCodeCheck.class);
     ruleMetadataLoader.addRulesByAnnotatedClass(repository, checks);
+
+    repository.rule("S100")
```
I think test is missing for default value

christophe-zurn-sonarsource replied:
Added test for default value of this rule in Kotlin

christophe-zurn-sonarsource commented:
@saberduck I've updated PR following your comments. I squashed the changes with their respective original commits so that we can keep the history on master when merging
